pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -0,0 +1,866 @@
+	}
+
+	if len(missingEntries) > 0 {
+		return "", "", fmt.Errorf("secret %s/%s contains the following empty TLS data entries: %s",
+			namespace, secretName, strings.Join(missingEntries, ", "))
+	}
+
+	return cert, key, nil
+}
+
+// loadServices is generating a WRR service, even when there is only one target.
+func loadServices(client Client, namespace string, targets []v1alpha1.HTTPRouteForwardTo) (*dynamic.Service, map[string]*dynamic.Service, error) {
+	services := map[string]*dynamic.Service{}
+
+	wrrSvc := &dynamic.Service{
+		Weighted: &dynamic.WeightedRoundRobin{
+			Services: []dynamic.WRRService{},
+		},
+	}
+
+	for _, forwardTo := range targets {
+		svc := dynamic.Service{
+			LoadBalancer: &dynamic.ServersLoadBalancer{
+				PassHostHeader: func(v bool) *bool { return &v }(true),
+			},
+		}
+
+		// TODO Handle BackendRefs
+
+		if forwardTo.ServiceName != nil {
+			service, exists, err := client.GetService(namespace, *forwardTo.ServiceName)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if !exists {
+				return nil, nil, errors.New("service not found")
+			}
+
+			if len(service.Spec.Ports) > 1 && forwardTo.Port == 0 {
+				// If the port is unspecified and the backend is a Service
+				// object consisting of multiple port definitions, the route
+				// must be dropped from the Gateway. The controller should
+				// raise the "ResolvedRefs" condition on the Gateway with the
+				// "DroppedRoutes" reason.  The gateway status for this route
+				// should be updated with a condition that describes the error
+				// more specifically.
+				log.WithoutContext().Errorf("A multiple ports Kubernetes Service cannot be used if unspecified forwardTo.Port")
+				continue
+			}
+
+			var portName string
+			var portSpec corev1.ServicePort
+			var match bool
+
+			for _, p := range service.Spec.Ports {
+				if forwardTo.Port == 0 || p.Port == int32(forwardTo.Port) {
+					portName = p.Name
+					portSpec = p
+					match = true
+					break
+				}
+			}
+
+			if !match {
+				return nil, nil, errors.New("service port not found")
+			}
+
+			endpoints, endpointsExists, endpointsErr := client.GetEndpoints(namespace, *forwardTo.ServiceName)
+			if endpointsErr != nil {
+				return nil, nil, endpointsErr
+			}
+
+			if !endpointsExists {
+				return nil, nil, errors.New("endpoints not found")
+			}
+
+			if len(endpoints.Subsets) == 0 {
+				return nil, nil, errors.New("subset not found")
+			}
+
+			var port int32
+			for _, subset := range endpoints.Subsets {
+				for _, p := range subset.Ports {
+					if portName == p.Name {
+						port = p.Port
+						break
+					}
+				}
+
+				if port == 0 {
+					return nil, nil, errors.New("cannot define a port")
+				}
+
+				protocol := getProtocol(portSpec, portName)
+
+				for _, addr := range subset.Addresses {
+					svc.LoadBalancer.Servers = append(svc.LoadBalancer.Servers, dynamic.Server{
+						URL: fmt.Sprintf("%s://%s:%d", protocol, addr.IP, port),
```

Comment:
Could you concat the IP and Port using `net.JoinHostPort`. Otherwise the URL will be invalid for IPv6 as it's expected to be `http://[2001::]:80`.